### PR TITLE
Changes for #465 and  #576

### PR DESCRIPTION
**Pull request by rikvdkleij**
Changes for #465 and #576

Do NOT merge these changes. These are changes made during Munihac and are not finished. PR is to share code with @arianvp  and @fizruk

### CODE REVIEW

**Comment by arianvp** [comment]
To give context to the rest. We added #465 and in the meanwhile discovered a lot of funny type errors for #576.

The result is, some instances are still missing for the auth combinators and should be implemented.

thanks a bunch @rikvdkleij !

File: servant-client/test/Servant/ClientSpec.hs [hunk 1]
```diff
@@ -218,7 +218,10 @@ type GenAuthAPI =
 genAuthAPI :: Proxy GenAuthAPI
 genAuthAPI = Proxy
 
-type instance AuthServerData (AuthProtect "auth-tag") = ()
+instance HasAuthServerData (AuthProtect "auth-tag") where
+  AuthServerData (AuthProtect "auth-tag") = ()
+
+-- type instance AuthServerData (AuthProtect "auth-tag") = ()
```
Comment (jkarni) [inline servant-client/test/Servant/ClientSpec.hs]: Remove?

File: servant-server/src/Servant/Server/Experimental/Auth.hs [hunk 2]
```diff
@@ -31,13 +32,21 @@ import           Servant.Server.Internal.RoutingApplication (addAuthCheck,
...
+import           GHC.TypeLits
+
 -- * General Auth
 
 -- | Specify the type of data returned after we've authenticated a request.
 -- quite often this is some `User` datatype.
 --
 -- NOTE: THIS API IS EXPERIMENTAL AND SUBJECT TO CHANGE
-type family AuthServerData a :: *
+-- type family AuthServerData a :: *
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Experimental/Auth.hs]: Remove?

File: servant-server/test/Servant/ServerSpec.hs [hunk 3]
```diff
@@ -71,8 +71,20 @@ import           Servant.Server.Internal.Context
...
+authCheck = BasicAuthCheck check
+  where
+    check _  = return (Authorized "servant")
+
+authHandler :: AuthHandler Request String
+authHandler =
+  let handler req = return "foo"
+  in mkAuthHandler handler
+
+-- type instance AuthServerData (AuthProtect "foo") = String
```
Comment (jkarni) [inline servant-server/test/Servant/ServerSpec.hs]: Remove

File: servant-server/src/Servant/Server/Experimental/Auth.hs [hunk 4]
```diff
@@ -31,13 +32,21 @@ import           Servant.Server.Internal.RoutingApplication (addAuthCheck,
...
+class HasAuthServerData a where
+
+type AuthServerData a :: *
+
+instance {-# OVERLAPPING #-} HasAuthServerData a where
+   type AuthServerData a = TypeError ('Text "aha this should never occur")
```
Comment (jkarni) [inline servant-server/src/Servant/Server/Experimental/Auth.hs]: Presumably CPP for GHC < 8, and a type error message? Could this be tested with doctest?

**Comment by arianvp** [comment]
I'll make some adjustments to this PR to make it a good cannidate for https://github.com/haskell-servant/servant/issues/465

**Comment by phadej** [comment]
@arianvp I assume you are now working on this?

**Comment by arianvp** [comment]
Correct.  I'll try to finish it by the end of the week

**Comment by phadej** [comment]
@arianvp Will your changes be breaking? If they aren't I could start testing whether 0.10 is ready for release

**Comment by arianvp** [comment]
No they should not introduce any backwards incompatible features.  We could
let it later in a next minor release safely. Don't feel held back to start
looking at readiness for 0.10

**Comment by phadej** [comment]
ping @arianvp

**Comment by phadej** [comment]
I guess we don't want this anymore if we are moving towards `servant-auth`?

**Comment by arianvp** [comment]
Yes I think we can close this. However, we should check if `servant-auth` _does_ have all the instances. If I recall correctly it doesn't have `-foreign` and `-docs` innstances currently.

**Comment by phadej** [comment]
Thanks @rikvdkleij for your contribution, unfortunately time have passed :/

**Pull request closed without merging**